File: app/src/main/java/uk/cybertecpro/infosecfeed/AlertPolicy.kt
package uk.cybertecpro.infosecfeed

enum class AlertMode {
    OFF,
    KEV_ONLY,
    KEV_AND_CRITICAL,
}

/** Pure selection policy: alert only on genuinely new, high-signal feed entries. */
object AlertPolicy {
    fun newAlerts(
        items: List<FeedItem>,
        seenIds: Set<String>,
        mode: AlertMode,
        limit: Int = 5,
    ): List<FeedItem> =
        items.asSequence()
            .filter { accepts(it, mode) && it.id !in seenIds }
            .sortedWith(
                compareByDescending<FeedItem> { it.severity == "EXPLOITED" }
                    .thenByDescending { it.published },
            )
            .take(limit)
            .toList()

    fun alertIds(items: List<FeedItem>, mode: AlertMode): Set<String> =
        items.asSequence()
            .filter { accepts(it, mode) }
            .map { it.id }
            .toSet()

    private fun accepts(item: FeedItem, mode: AlertMode): Boolean = when (mode) {
        AlertMode.OFF -> false
        AlertMode.KEV_ONLY -> item.severity == "EXPLOITED"
        AlertMode.KEV_AND_CRITICAL -> item.severity == "EXPLOITED" || item.severity == "CRITICAL"
    }
}

File: windows/InfoSecFeed.Windows/Services/AlertPolicy.cs
using InfoSecFeed.Windows.Models;

namespace InfoSecFeed.Windows.Services;

public enum AlertMode
{
    Off,
    KevOnly,
    KevAndCritical,
}

public static class AlertPolicy
{
    public static IReadOnlyList<FeedItem> SelectNew(
        IEnumerable<FeedItem> items,
        IReadOnlySet<string> seenIds,
        AlertMode mode,
        int limit = 5) =>
        items
            .Where(item => Accepts(item, mode) && !seenIds.Contains(item.Id))
            .OrderByDescending(item => item.Severity == "EXPLOITED")
            .ThenByDescending(item => item.Published)
            .Take(limit)
            .ToList();

    public static HashSet<string> AlertIds(IEnumerable<FeedItem> items, AlertMode mode) =>
        items.Where(item => Accepts(item, mode)).Select(item => item.Id).ToHashSet();

    private static bool Accepts(FeedItem item, AlertMode mode) => mode switch
    {
        AlertMode.Off => false,
        AlertMode.KevOnly => item.Severity == "EXPLOITED",
        AlertMode.KevAndCritical => item.Severity is "EXPLOITED" or "CRITICAL",
        _ => false,
    };
}
